File: core/noise.js
import { lerp } from '../js/utils.js';

function hash2(x, y, s = 1337) {
  let h = Math.imul(x, 374761393) ^ Math.imul(y, 668265263) ^ Math.imul(s, 362437);
  h = (h ^ (h >>> 13)) >>> 0;
  h = Math.imul(h, 1274126177) >>> 0;
  h ^= h >>> 16;
  return (h >>> 0) / 4294967295;
}

export function valueNoise2D(x, y, seed = 0) {
  const x0 = Math.floor(x), y0 = Math.floor(y);
  const tx = x - x0, ty = y - y0;
  const a = hash2(x0, y0, seed), b = hash2(x0 + 1, y0, seed);
  const c = hash2(x0, y0 + 1, seed), d = hash2(x0 + 1, y0 + 1, seed);
  const u = tx * tx * (3 - 2 * tx), v = ty * ty * (3 - 2 * ty);
  return lerp(lerp(a, b, u), lerp(c, d, u), v);
}

export function fbm2D(x, y, octaves = 5, lacunarity = 2, gain = 0.5, seed = 0) {
  let amp = 1, freq = 1, sum = 0, ampSum = 0;
  for (let i = 0; i < octaves; i++) {
    sum += amp * valueNoise2D(x * freq, y * freq, seed + i * 17);
    ampSum += amp;
    freq *= lacunarity;
    amp *= gain;
  }
  return ampSum > 0 ? sum / ampSum : 0;
}

File: Core/Noise.cs
using System.Numerics;

namespace HeightmapMinecraft.Core;

public static class Noise
{
    private static float Fade(float t) => t * t * (3f - 2f * t);

    public static float[] ValueNoise2D(int size, float frequency, int seed)
    {
        var grid = (int)MathF.Ceiling(size * frequency) + 3;
        var random = new Random(seed);
        var lattice = new float[grid * grid];
        for (var i = 0; i < lattice.Length; i++)
            lattice[i] = (float)random.NextDouble();

        var output = new float[size * size];
        for (var y = 0; y < size; y++)
        {
            var fy = y * frequency;
            var y0 = (int)MathF.Floor(fy);
            var y1 = y0 + 1;
            var ty = Fade(fy - y0);

            for (var x = 0; x < size; x++)
            {
                var fx = x * frequency;
                var x0 = (int)MathF.Floor(fx);
                var x1 = x0 + 1;
                var tx = Fade(fx - x0);

                var v00 = lattice[y0 * grid + x0];
                var v10 = lattice[y0 * grid + x1];
                var v01 = lattice[y1 * grid + x0];
                var v11 = lattice[y1 * grid + x1];

                var a = v00 * (1f - tx) + v10 * tx;
                var b = v01 * (1f - tx) + v11 * tx;
                output[y * size + x] = a * (1f - ty) + b * ty;
            }
        }

        return output;
    }

    public static float[] Fbm(int size, float baseFreq, int octaves, float lacunarity, float gain, int seed)
    {
        var outArr = new float[size * size];
        var amp = 1f;
        var freq = baseFreq;
        var ampSum = 0f;

        for (var i = 0; i < octaves; i++)
        {
            var n = ValueNoise2D(size, freq, seed + i * 101);
            for (var p = 0; p < outArr.Length; p++)
                outArr[p] += amp * (n[p] * 2f - 1f);

            ampSum += amp;
            amp *= gain;
            freq *= lacunarity;
        }

        for (var p = 0; p < outArr.Length; p++)
            outArr[p] = (outArr[p] / ampSum + 1f) * 0.5f;

        return outArr;
    }

    public static float[] RidgedNoise(int size, float baseFreq, int octaves, int seed)
    {
        var n = Fbm(size, baseFreq, octaves, 2f, 0.56f, seed);
        for (var i = 0; i < n.Length; i++)
        {
            var r = 1f - MathF.Abs(2f * n[i] - 1f);
            n[i] = Math.Clamp(r, 0f, 1f);
        }

        return n;
    }
}
